zookeeper-server/src/main/java/org/apache/zookeeper/Login.java

```diff
@@ -120,172 +272,36 @@ public Login(final String loginContextName, CallbackHandler callbackHandler, fin
-                                Shell.execCommand(cmd, kinitArgs);
-                                break;
-                            } catch (Exception e) {
-                                if (retry > 0) {
-                                    --retry;
-                                    // sleep for 10 seconds
-                                    try {
-                                        Thread.sleep(10 * 1000);
-                                    } catch (InterruptedException ie) {
-                                        LOG.error("Interrupted while renewing TGT, exiting Login thread");
-                                        return;
-                                    }
-                                } else {
-                                    LOG.warn(
-                                        "Could not renew TGT due to problem running shell command: '{} {}'."
-                                            + " Exiting refresh thread.",
-                                        cmd,
-                                        kinitArgs,
-                                        e);
-                                    return;
-                                }
-                            }
-                        }
-                    }
-                    try {
-                        int retry = 1;
-                        while (retry >= 0) {
-                            try {
-                                reLogin();
-                                break;
-                            } catch (LoginException le) {
-                                if (retry > 0) {
-                                    --retry;
-                                    // sleep for 10 seconds.
-                                    try {
-                                        Thread.sleep(10 * 1000);
-                                    } catch (InterruptedException e) {
-                                        LOG.error("Interrupted during login retry after LoginException:", le);
-                                        throw le;
-                                    }
-                                } else {
-                                    LOG.error("Could not refresh TGT for principal: {}.", principal, le);
-                                }
-                            }
-                        }
-                    } catch (LoginException le) {
-                        LOG.error("Failed to refresh TGT: refresh thread exiting now.", le);
-                        break;
-                    }
-                }
-            }
-        });
-        t.setDaemon(true);
+        startLoginThread();
     }
 
-    public void startThreadIfNeeded() {
-        // thread object 't' will be null if a refresh thread is not needed.
-        if (t != null) {
+    private void startLoginThread() {
+        // Start the thread only if not canceled before reaching here.
+        synchronized (loginThreadLock) {
```

Comment:
I think in Java you can use synchronized methods if the block encapsulates the entire method. You don't need to introduce a separate lock object and there's no performance hit either. I believe the syntax sugar uses the same locking method, but on the object itself.

Reply:
@anmolnar 
Yes, I used synchronized block initially, but realized there are already a bunch of synchronized methods in Login (for ex reLogin() which does a krb relogin). We should be able to cancel/interrupt the thread even if it's in the process of relogin. So, that's why I turned to a synchronized block.

Reply:
Got you.